### ISSUE TITLE
Make source optional behind Cargo features, or abstact them behind a trait and remove them entirely

### ISSUE THREAD

**Issue by djeedai** [mehcode/config-rs#320]
Currently it seems that the various sources (env, file, net) of config are all mandatory, and force a long list of mandatory dependencies. It would be great if at least those sources would be behind feature flags, to allow consuming crates/projects to tailor it to their use case without requiring that many dependencies.

But pushing that idea even further, if I look at the crate I think I see 2 separate roles:
1. the core responsibility for this crate is to manage configs, serializing and deserializing them, and managing overriding and priorities among sources;
2. it also attempts to abstract network or filesystem or environment variable access to provide an out-of-the-box experience, but in doing so limits the flexibility and extensibility to those built-in sources and the set of features they support.

It would be much better in my opinion to abstract the sources under a trait that the users can implement as they want, and then just manage a collection of abstract sources, without the need to specialize any of them in this crate. This would delegate the actual process of reading and writing serialized data to other crates, separating the responsibilities, and providing a lot more flexibility and extensibility. I suspect this would also simplify sources management quite a bit, since they'd be manipulated via a common trait.

**Comment by matthiasbeyer** [comment]
Hi, thanks for opening this issue :heart: (and [writing this comment](https://github.com/mehcode/config-rs/issues/111#issuecomment-1106832438)). I have been thinking a lot about how to go on with the development of this crate lately.

I totally agree. In my opinion, we should even go one step further and re-think the whole crate. I agree that the core responsibility for this crate is the "overwrite" functionality, where one can layer configurations on each other.

I am thinking on how to tackle this specific issue but also the whole idea of re-thinking the crate. I might open a "project" for this crate and start filling it with ideas (one idea per issue) and then see what we can collect and then agree on.

**Comment by epage** [comment]
In terms of file formats, they are currently behind features but enabled by default.  In #621, we'll disable them by default.

In terms of sources, the required dependency is `pathdiff`

While I think its important to allow third-parties to extend `config` without having to bake everything in, moving everything out would also have an ergonomics issue.  I lean towards closing this issue but if there is a reason for us to re-evaluate, let us know!

**Comment by djeedai** [comment]
As I recall, the issue for me was about security and making sure some sources are inaccessible. Not all types of apps are happy letting users override settings from environment variables for example. So having less dependencies is not necessarily about building faster but more about not pulling into your codebases crates for features you actively don't want.

**Comment by djeedai** [comment]
Yeah actually I've just re-read my comment (linked above) and that's exactly what I was saying at the time, mandatory networking dependency is a deal breaker for security reasons.

**Comment by epage** [comment]
At this point [there is no mandatory network dependency](https://crates.io/crates/config/0.15.18/dependencies).

EDIT: [That was also true in 2022](https://crates.io/crates/config/0.13.1/dependencies)
